File: front/lib/models/goal.dart
/// 학습 목표 모델
class Goal {
  final int id;
  final String title;
  final DateTime? startDate;
  final DateTime? endDate;
  final int totalTargetAmount;
  final String targetUnit;
  final int currentProgress;
  final bool isCompleted;
  final DateTime? completedAt;

  Goal({
    required this.id,
    required this.title,
    this.startDate,
    this.endDate,
    required this.totalTargetAmount,
    required this.targetUnit,
    required this.currentProgress,
    required this.isCompleted,
    this.completedAt,
  });

  factory Goal.fromJson(Map<String, dynamic> json) {
    return Goal(
      id: json['id'] is int ? json['id'] : int.parse(json['id'].toString()),
      title: json['title'] ?? '',
      startDate: json['startDate'] != null
          ? DateTime.parse(json['startDate'])
          : null,
      endDate: json['endDate'] != null
          ? DateTime.parse(json['endDate'])
          : null,
      totalTargetAmount: json['totalTargetAmount'] ?? 0,
      targetUnit: json['targetUnit'] ?? '',
      currentProgress: json['currentProgress'] ?? 0,
      isCompleted: json['isCompleted'] ?? false,
      completedAt: json['completedAt'] != null
          ? DateTime.parse(json['completedAt'])
          : null,
    );
  }

  /// 진행률 (%)
  double get progressPercentage {
    if (totalTargetAmount == 0) return 0;
    return (currentProgress / totalTargetAmount * 100).clamp(0, 100);
  }

  /// 남은 목표량
  int get remainingAmount {
    return (totalTargetAmount - currentProgress).clamp(0, totalTargetAmount);
  }
}

File: backEnd/learnkit-backend/src/main/java/com/learnkit/backend/domain/Goal.java
package com.learnkit.backend.domain;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Getter
@Entity
@Table(name = "goals")
@NoArgsConstructor // 파라미터가 없는 기본 생성자. JPA가 DB에서 데이터를 가져오기 위한 빈 박스 생성.
public class Goal extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @ManyToOne(fetch = FetchType.LAZY) // N:1 관계, 지연로딩 방식
    @JoinColumn(name = "user_id", nullable = false) // 컬럼 매핑
    private User user;

    @Column(nullable = false)
    private String title;

    private LocalDate startDate;

    private LocalDate endDate;

    @Column(nullable = false)
    private int totalTargetAmount;

    @Column(nullable = false, length = 50) // 더 줄일만함
    private String targetUnit;

    @Column(nullable = false)
    private int currentProgress = 0;

    @Column(nullable = false)
    private boolean isCompleted = false;

    private LocalDateTime completedAt;

    // 생성자
    public Goal(String title, LocalDate startDate, LocalDate endDate, int totalTargetAmount, String targetUnit) {
        this.title = title;
        this.startDate = startDate;
        this.endDate = endDate;
        this.totalTargetAmount = totalTargetAmount;
        this.targetUnit = targetUnit;
    }

    public void setUser(User user) {
        this.user = user;
    }

    // 진행도 업데이트
    public void addProgress(int amount) {
        this.currentProgress += amount; // 학습 종료 후 사용자가 입력

        // 이번 학습으로 목표치 달성 시
        if (this.currentProgress >= this.totalTargetAmount && !this.isCompleted) {
            this.isCompleted = true;
            this.completedAt = LocalDateTime.now();
        }
    }

    // 학습 목표 수정
    public void update(String title, LocalDate startDate, LocalDate endDate, Integer totalTargetAmount, String targetUnit){

        // 입력한 필드만 수정
        if (title != null) this.title = title;
        if (startDate != null) this.startDate = startDate;
        if (endDate != null) this.endDate = endDate;
        if (totalTargetAmount != null) this.totalTargetAmount = totalTargetAmount;
        if (targetUnit != null) this.targetUnit = targetUnit;

    }

}
